source/common/callback/callback.h

```diff
@@ -0,0 +1,175 @@
+ * case where the event consumer outlives its producers.
+ *
+ * Caller and Receiver provide some additional safety. A Receiver owns a callback function, and
+ * can produce Callers which function as weak references to it. The receiver's callback function
+ * can only be invoked via its callers. If the receiver is destroyed, invoking its callers has
+ * no effect, so none of the callback's captures can be unsafely dereferenced.
+ *
+ * When implementing this pattern, an event consumer would own a Receiver and an event producer
+ * would own the corresponding Caller. For example:
+ *
+ *   using EventCaller = Callback::Caller<const Event&>;
+ *   using EventReceiver = Callback::Receiver<const Event&>;
+ *
+ *   class MyEventConsumer {
+ *   public:
+ *     MyEventConsumer(EventProducer& producer) {
+ *       producer.attach(receiver_.caller());
+ *     }
+ *   private:
+ *     EventReceiver receiver_([this]() { ... handle event ... });
+ *   };
+ *
+ *   class EventProducer {
+ *   public:
+ *     void attach(EventCaller caller) { caller_ = caller; }
+ *     void invoke() { caller_(... some event ... ); }
+ *   private:
+ *     EventCaller caller_;
+ *   };
+ */
+namespace Callback {
+
+// Forward-declaration for Caller's friend declaration.
+template <typename... Args> class Receiver;
+
+/**
+ * Caller: simple wrapper for a weak_ptr to a callback function. Copyable and movable.
+ */
+template <typename... Args> class Caller {
+public:
+  /**
+   * Default constructor for default / value initialization.
+   */
+  Caller() = default;
+
+  /**
+   * Implicit conversion to bool, to test whether the corresponding Receiver is still available.
+   * @return true if the corresponding Receiver is still available, false otherwise.
+   */
+  operator bool() const { return !fn_.expired(); }
+
+  /**
+   * Invoke the corresponding Receiver's callback, if it is still available. If the receiver has
+   * been destroyed or reset, this has no effect.
+   * @param args the arguments, if any, to pass to the receiver's callback function.
+   */
+  void operator()(Args... args) const {
+    auto locked_fn(fn_.lock());
+    if (locked_fn) {
+      (*locked_fn)(args...);
+    }
+  }
+
+private:
+  /**
+   * Can only be constructed by a Receiver
+   */
+  friend Receiver<Args...>;
+  Caller(std::weak_ptr<std::function<void(Args...)>> fn) : fn_(std::move(fn)) {}
+
+  std::weak_ptr<std::function<void(Args...)>> fn_;
+};
+
+/**
+ * Receiver: simple wrapper for a shared_ptr to a callback function. Copyable and movable, but
+ * typically should be owned uniquely by the owner of any pointers and references captured by its
+ * handler function. For example, if `this` is captured by the handler function, `this` should
+ * probably also own the Receiver.
+ */
+template <typename... Args> class Receiver {
+public:
+  /**
+   * Default constructor for default / value initialization.
+   */
+  Receiver() = default;
```

Comment:
Ditto, what happens if this isn't here?

Reply:
`Receiver` is meant to have similar behavior to both `std::function<...>` and `std::shared_ptr<...>`. You can initialize either of those to be empty, and you can reset them to an empty state after construction, and I figure it's worth preserving that ability here.